#### V2IDA
![alt tag](https://user-images.githubusercontent.com/57667417/69444736-bfc1e200-0d2f-11ea-94f4-592adbb3b0ed.jpeg)

# Viral Vaccine genetic Diversity Analyzer

This script demonstrates using shell scripting to provide an automatized and user-friendly scientific pipeline to 


To this end, a pipeline has been developed to allow researchers at the CGSB to rapidly identify and annotate variants.


QuasiRecomb
Can perform region

### List of Tools Used in this Pipeline

[BWA](https://github.com/lh3/bwa)

[Samtools](https://github.com/samtools/samtools)

[Picard](https://github.com/broadinstitute/picard)

[GATK](https://github.com/broadinstitute/gatk)

[QuasiRecomb](https://github.com/cbg-ethz/QuasiRecomb)


All requirements should be downloaded and installed by the user. 
As default the tools should be in path.
Or in the 00_Exe folder

### Overview

![alt tag](https://user-images.githubusercontent.com/57667417/69445815-dd904680-0d31-11ea-8885-a2c03c968c92.png)


\ 
### Walk thourgh
#### Step 01 - User input
#### Step 02 - Alignment
#### Step 03 - Variant Calling

The pipeline employs the Genome Analysis Toolkit (GATK) to perform variant calling and is based on the best practices for variant discovery analysis outlined by the Broad Institute. 


#### Step 04 - Quasispecies reconstruction

QuasiRecomb


### Citation 
